### PR TITLE
[IR] accommodate preIR and IR deployments side by side

### PR DESCRIPTION
**Pull request by janelletavares**
## Description of change

Fixes https://github.com/meroxa/product/issues/544

## Type of change

- [x]  New feature
- [ ]  Bug fix
- [ ]  Refactor
- [ ]  Documentation

## How was this tested?

- [x]  Unit Tests
- [ ]  Tested in staging
- [ ]  Tested in minikube

## Demo

| before | after |
|--------|-------|
|||

### CODE REVIEW

File: cmd/meroxa/root/apps/deploy.go [hunk 1]
```diff
@@ -649,25 +638,54 @@ func (d *Deploy) Execute(ctx context.Context) error {
...
 		return err
 	}
 
-	return d.deployApp(ctx, d.fnName, gitSha, d.specVersion)
+	if d.specVersion == "" {
+		app, err = d.checkPreIRApplication(ctx)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (janelletavares) [inline cmd/meroxa/root/apps/deploy.go:684]: @TODO get the latest deployment status, potentially in a loop if it's not implemented as a stream

File: cmd/meroxa/root/apps/deploy.go [hunk 2]
```diff
@@ -649,25 +638,54 @@ func (d *Deploy) Execute(ctx context.Context) error {
...
-	if err := d.tearDownExistingResources(ctx); err != nil {
-		return err
+	var app *meroxa.Application
+	if d.specVersion == "" {
+		// ⚠️ This is only until we re-deploy applications applying only the changes made
+		if err := d.tearDownExistingResources(ctx); err != nil {
+			return err
+		}
+	} else {
+		// Intermediate Representation Workflow
```
Comment (janelletavares) [inline cmd/meroxa/root/apps/deploy.go:653]: @TODO check feature flag

**Comment by janelletavares** [comment]
I tested the pre IR and IR code paths for go apps.  Test results pending for py and js.

File: cmd/meroxa/root/apps/deploy.go [hunk 3]
```diff
@@ -238,22 +242,24 @@ func (d *Deploy) getPlatformImage(ctx context.Context) (string, error) {
...
-	if err != nil {
-		return err
+	if specVersion != "" {
+		input := &meroxa.CreateDeploymentInput{
+			Application: meroxa.EntityIdentifier{Name: null.StringFrom(d.appName)},
+			GitSha:      gitSha,
+			SpecVersion: null.StringFrom(specVersion),
+			Spec:        null.StringFrom(deploymentSpec),
+		}
+		_, err = d.client.CreateDeployment(ctx, input)
```
Comment (raulb) [inline cmd/meroxa/root/apps/deploy.go:256]: I thought this was going to be done from the API. 

Are we planning on making the API create the deployment in some point? I believe it should be more stable to leave the API the responsibility of doing that.

File: cmd/meroxa/root/apps/deploy.go [hunk 4]
```diff
@@ -600,7 +606,45 @@ func (d *Deploy) prepareAppName(ctx context.Context) string {
 	return appName
 }
 
+func (d *Deploy) waitForDeployment(ctx context.Context) error {
+	startTime := time.Now().UTC()
+	logs := []string{}
+	for time.Now().UTC().Sub(startTime) < 5*time.Minute {
```
Comment (raulb) [inline cmd/meroxa/root/apps/deploy.go:612]: Is this running every ms or how often?

File: cmd/meroxa/root/apps/deploy.go (hunk 3 above)
Comment (lyuboxa) [reply cmd/meroxa/root/apps/deploy.go:256]: I thought creating the deployment which carries the spec from the CLI was the way? Since this is what carries the IR?

File: cmd/meroxa/root/apps/deploy.go [hunk 5]
```diff
@@ -600,7 +606,45 @@ func (d *Deploy) prepareAppName(ctx context.Context) string {
...
+		var deployment *meroxa.Deployment
+		deployment, err := d.client.GetLatestDeployment(ctx, d.appName)
+		if err != nil {
+			return err
+		}
+		newLogs := strings.Split(deployment.OutputLog.String, "\n")
```
Comment (lyuboxa) [inline cmd/meroxa/root/apps/deploy.go:618]: What is output log? I see it in meroxa-go log but the platform doesn't do anything with that. I'd rather we don't use this until it is implemented

File: cmd/meroxa/root/apps/deploy.go (hunk 4 above)
Comment (lyuboxa) [reply cmd/meroxa/root/apps/deploy.go:612]: can we use a ticket with a context timeout instead?
kind of looks like this
```
	cctx, cancel := context.WithTimeout(ctx, <timeout>)
	defer cancel()

	t := time.NewTicker(1 * time.Second)
	defer t.Stop()

	for {
		select {
		case <-t.C:
			<do work>
		case <-cctx.Done():
                       <timeout expired>
			return cctx.Err()
		}
	}
```

File: cmd/meroxa/root/apps/deploy.go (hunk 3 above)
Comment (janelletavares) [reply cmd/meroxa/root/apps/deploy.go:256]: I don't understand the comment, Raul.  Could you rephrase?  What exactly did you think was going to be done in the API?  Did you think the application endpoint would create a populated deployment?  The Application endpoint will create an empty pipeline in v2, and the v1 application endpoint backfills an empty deployment for already-populated applications.

File: cmd/meroxa/root/apps/deploy.go (hunk 4 above)
Comment (janelletavares) [reply cmd/meroxa/root/apps/deploy.go:612]: yeah, it doesn't sleep or backoff at all, Raul, a small sleep might be good.  Given that deployments are immediately done with no output logs, this doesn't do much of anything now.  mostly exercised by the unit test.

File: cmd/meroxa/root/apps/deploy.go (hunk 5 above)
Comment (janelletavares) [reply cmd/meroxa/root/apps/deploy.go:618]: the worker doesn't add any outlogs right now, but in the future i would expect one log per step. we want the logs to roughly match to the CLI logs today

**Review by jayjayjpg — APPROVED** [review]
This reads great, thank you for putting all of this together so quickly! :sparkles:

File: cmd/meroxa/root/apps/deploy.go (hunk 3 above)
Comment (janelletavares) [reply cmd/meroxa/root/apps/deploy.go:258]: this is the result of the second RFC.  this is what we do if we make sure the application exists before the deployment.  we create deployments and not applications after the application exists.
